core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -126,110 +122,75 @@ protected boolean resolveIndex() {
+                Engine.IndexResult indexResult = executeIndexRequestOnPrimary(indexRequest, primary, mappingUpdatedAction);
+                response = indexResult.hasFailure() ? null :
+                        new IndexResponse(primary.shardId(), indexRequest.type(), indexRequest.id(), indexResult.getSeqNo(),
+                                indexResult.getVersion(), indexResult.isCreated());
+                operationResult = indexResult;
+                replicaRequest = request.items()[requestIndex];
+                break;
+            case UPDATE:
+                UpdateResultHolder updateResultHolder = executeUpdateRequest(((UpdateRequest) itemRequest),
+                    primary, metaData, request, requestIndex);
+                operationResult = updateResultHolder.operationResult;
+                response = updateResultHolder.response;
+                replicaRequest = updateResultHolder.replicaRequest;
+                break;
+            case DELETE:
+                final DeleteRequest deleteRequest = (DeleteRequest) itemRequest;
+                Engine.DeleteResult deleteResult = executeDeleteRequestOnPrimary(deleteRequest, primary);
+                response = deleteResult.hasFailure() ? null :
+                        new DeleteResponse(request.shardId(), deleteRequest.type(), deleteRequest.id(), deleteResult.getSeqNo(),
+                                deleteResult.getVersion(), deleteResult.isFound());
+                operationResult = deleteResult;
+                replicaRequest = request.items()[requestIndex];
+                break;
+            default: throw new IllegalStateException("unexpected opType [" + itemRequest.opType() + "] found");
+        }
 
-            // update the bulk item request because update request execution can mutate the bulk item request
-            request.items()[requestIndex] = replicaRequest;
-            if (operationResult == null) { // in case of noop update operation
-                assert response.getResult() == DocWriteResponse.Result.NOOP
-                    : "only noop update can have null operation";
-                replicaRequest.setIgnoreOnReplica();
-                replicaRequest.setPrimaryResponse(new BulkItemResponse(replicaRequest.id(), opType, response));
-            } else if (operationResult.hasFailure() == false) {
-                location = locationToSync(location, operationResult.getTranslogLocation());
-                BulkItemResponse primaryResponse = new BulkItemResponse(replicaRequest.id(), opType, response);
-                replicaRequest.setPrimaryResponse(primaryResponse);
-                // set the ShardInfo to 0 so we can safely send it to the replicas. We won't use it in the real response though.
-                primaryResponse.getResponse().setShardInfo(new ShardInfo());
+        // update the bulk item request because update request execution can mutate the bulk item request
+        request.items()[requestIndex] = replicaRequest;
```

Comment:
yeah, the problem is the update which is much trickier. In a sense, it's an update once kind of logic. Cleaning this is a fight for another day..